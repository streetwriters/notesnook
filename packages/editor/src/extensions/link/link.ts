/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import {
  Command,
  Mark,
  getMarkType,
  isMarkActive,
  markInputRule,
  markPasteRule,
  mergeAttributes,
  PasteRuleMatch
} from "@tiptap/core";
import { Plugin, TextSelection } from "@tiptap/pm/state";
import { find, registerCustomProtocol, reset } from "linkifyjs";
import { autolink } from "./helpers/autolink.js";
import { clickHandler } from "./helpers/clickHandler.js";
import { pasteHandler } from "./helpers/pasteHandler.js";
import { ImageNode } from "../image/index.js";
import { selectionToOffset } from "../../utils/prosemirror.js";

export interface LinkProtocolOptions {
  scheme: string;
  optionalSlashes?: boolean;
}

export interface LinkOptions {
  /**
   * If enabled, it adds links as you type.
   */
  autolink: boolean;
  /**
   * An array of custom protocols to be registered with linkifyjs.
   */
  protocols: Array<LinkProtocolOptions | string>;
  /**
   * If enabled, links will be opened on click.
   */
  openOnClick: boolean;
  /**
   * Adds a link to the current selection if the pasted content only contains an url.
   */
  linkOnPaste: boolean;
  /**
   * A list of HTML attributes to be rendered.
   */
  HTMLAttributes: Record<string, any>;
  /**
   * A validation function that modifies link verification for the auto linker.
   * @param url - The url to be validated.
   * @returns - True if the url is valid, false otherwise.
   */
  validate?: (url: string) => boolean;
}

export type LinkAttributes = {
  href: string;
  target?: string | null;
  rel?: string | null;
  class?: string | null;
  title?: string | null;
};
declare module "@tiptap/core" {
  interface Commands<ReturnType> {
    link: {
      /**
       * Set a link mark
       */
      setLink: (attributes: LinkAttributes) => ReturnType;
      /**
       * Toggle a link mark
       */
      toggleLink: (attributes: LinkAttributes) => ReturnType;
      /**
       * Unset a link mark
       */
      unsetLink: () => ReturnType;
    };
  }
}

export const linkRegex = /(?:__|[*#])|\[([^\]]+)\]\(.*?\)/gm;
const regExp = /(?:__|[*#])|\[.*?\]\((.*?)\)/gm;
export const Link = Mark.create<LinkOptions>({
  name: "link",

  priority: 1000,

  keepOnSplit: false,

  onCreate() {
    this.options.protocols.forEach((protocol) => {
      if (typeof protocol === "string") {
        registerCustomProtocol(protocol);
        return;
      }
      registerCustomProtocol(protocol.scheme, protocol.optionalSlashes);
    });
  },

  onDestroy() {
    reset();
  },

  inclusive() {
    return this.options.autolink;
  },

  addOptions() {
    return {
      openOnClick: true,
      linkOnPaste: true,
      autolink: true,
      protocols: [],
      HTMLAttributes: {
        target: "_blank",
        rel: "noopener noreferrer nofollow",
        class: null,
        title: null
      },
      validate: undefined
    };
  },

  addAttributes() {
    return {
      href: {
        default: null
      },
      target: {
        default: this.options.HTMLAttributes.target
      },
      rel: {
        default: this.options.HTMLAttributes.rel
      },
      class: {
        default: this.options.HTMLAttributes.class
      },
      title: {
        default: this.options.HTMLAttributes.title
      }
    };
  },

  parseHTML() {
    // return [{ tag: 'a[href]:not([href *= "javascript:" i])' }];
    return [{ tag: "a" }];
  },

  renderHTML({ HTMLAttributes }) {
    // False positive; we're explicitly checking for javascript: links to ignore them
    if (HTMLAttributes.href?.startsWith?.("javascript:")) {
      // strip out the href
      return [
        "a",
        mergeAttributes(this.options.HTMLAttributes, {
          ...HTMLAttributes,
          href: ""
        }),
        0
      ];
    }
    return [
      "a",
      mergeAttributes(this.options.HTMLAttributes, HTMLAttributes),
      0
    ];
  },

  addCommands() {
    return {
      setLink: (attributes) => insertLink(attributes, false),

      toggleLink: (attributes) => insertLink(attributes, true),

      unsetLink:
        () =>
        ({ chain }) => {
          return chain()
            .unsetMark(this.name, { extendEmptyMarkRange: true })
            .setMeta("preventAutolink", true)
            .run();
        }
    };
  },

  addInputRules() {
    return [
      ...(this.parent?.() || []),
      markInputRule({
        find: linkRegex,
        type: this.type,
        getAttributes: (match) => {
          return {
            href: regExp.exec(match[0])?.[1]
          };
        }
      })
    ];
  },

  addPasteRules() {
    return [
      markPasteRule({
        find: linkRegex,
        type: this.type,
        getAttributes(match) {
          // reset in case of multiple matches
          regExp.lastIndex = 0;
          return {
            href: regExp.exec(match[0])?.[1]
          };
        }
      })
    ];
  },

  addProseMirrorPlugins() {
    const plugins: Plugin[] = [];

    if (this.options.autolink) {
      plugins.push(
        autolink({
          type: this.type,
          validate: this.options.validate
        })
      );
    }

    if (this.options.openOnClick) {
      plugins.push(
        clickHandler({
          type: this.type,
          editor: this.editor
        })
      );
    }

    plugins.push(
      pasteHandler({
        editor: this.editor,
        type: this.type,
        linkOnPaste: this.options.linkOnPaste
      })
    );

    return plugins;
  },

  addKeyboardShortcuts() {
    return {
      Space: ({ editor }) => {
        const { from, to } = editor.state.selection;

        if (!editor.state.doc.rangeHasMark(from, to + 1, this.type)) {
          const { tr } = editor.state;
          tr.removeStoredMark(editor.schema.marks.link);
          editor.view.dispatch(tr);
        }

        return false;
      }
    };
  }
});

const insertLink: (attributes: LinkAttributes, toggle?: boolean) => Command =
  (attributes, toggle) =>
  ({ chain, editor }) => {
    let commandChain = chain();

    const offset = selectionToOffset(editor.state);
    if (!offset) return false;
    const { from, to, node } = offset;

    const isSelection = !editor.state.selection.empty;

    const isEditing = isMarkActive(editor.state, Link.name);
    const isImage = node?.type.name === ImageNode.name;

    if (isEditing && node) {
      if (!isImage) {
        const markType = getMarkType(Link.name, editor.schema);
        commandChain = commandChain.command(({ tr }) => {
          tr.removeMark(from, to, markType);
          tr.insertText(
            attributes.title || node.textContent,
            tr.mapping.map(from),
            tr.mapping.map(to)
          );
          tr.setSelection(
            TextSelection.create(
              tr.doc,
              tr.mapping.map(from),
              tr.mapping.map(to)
            )
          );
          return true;
        });
      }

      return commandChain
        .extendMarkRange("link")
        .setMark(Link.name, attributes)
        .command(({ tr }) => {
          tr.setSelection(
            TextSelection.create(
              tr.doc,
              tr.mapping.map(editor.state.selection.from),
              tr.mapping.map(editor.state.selection.to)
            )
          );
          return true;
        })
        .focus(undefined, { scrollIntoView: true })
        .run();
    }

    commandChain = toggle
      ? commandChain.toggleMark(Link.name, attributes, {
          extendEmptyMarkRange: true
        })
      : commandChain.extendMarkRange(Link.name).setMark(Link.name, attributes);

    if (!isImage)
      commandChain = commandChain.insertContent(
        attributes.title || attributes.href
      );

    if (!isSelection && !isImage) {
      commandChain = commandChain.command(({ tr }) => {
        tr.insertText(" ", tr.mapping.map(to));
        tr.removeMark(
          tr.mapping.map(to) - 1,
          tr.mapping.map(to),
          getMarkType(Link.name, editor.schema)
        );
        return true;
      });
    }

    return commandChain.focus().setMeta("preventAutolink", true).run();
  };
