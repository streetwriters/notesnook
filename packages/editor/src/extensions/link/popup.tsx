/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { showPopup } from "../../components/popup-presenter";
import { Button, Flex, Input, Text } from "@theme-ui/components";
import { PopupWrapper } from "../../components/popup-presenter";
import { Icon } from "@notesnook/ui";
import { useState } from "react";
import { Icons } from "../../toolbar";
import { getToolbarElement } from "../../toolbar/utils/dom";
import { useToolbarLocation } from "../../toolbar/stores/toolbar-store";
import { Editor } from "@tiptap/core";
import { TextSelection } from "prosemirror-state";

type LinkFloatingMenu = {
  onClose: () => void;
  isOpen: boolean;
  editor: Editor;
};

export function LinkFloatingMenu(props: LinkFloatingMenu) {
  const { onClose, isOpen, editor } = props;
  const toolbarLocation = useToolbarLocation();
  const isBottom = toolbarLocation === "bottom";
  const [open, setOpen] = useState(true);
  const [href, setHref] = useState("");
  const [text, setText] = useState("");

  const { from, to } = editor.state.selection;
  const selectedText = editor.state.doc.textBetween(from, to);

  return (
    <PopupWrapper
      group="link-floating-menu"
      id="link-floating-menu"
      isOpen={open}
      position={{
        isTargetAbsolute: true,
        target: getToolbarElement(),
        align: "center",
        location: isBottom ? "top" : "below",
        yOffset: 10
      }}
    >
      <Flex sx={{ p: 1, flexDirection: "column", width: ["auto", 250] }}>
        <Flex
          className="movable"
          sx={{
            justifyContent: "space-between",
            alignItems: "center",
            p: 2
          }}
        >
          <Text variant={"title"}>Insert Link</Text>
          <Button
            variant={"secondary"}
            sx={{ p: 0, bg: "transparent" }}
            onClick={() => setOpen(false)}
          >
            <Icon path={Icons.close} size={"big"} />
          </Button>
        </Flex>
        <Input
          type="text"
          placeholder="Link text"
          defaultValue={selectedText}
          sx={{ mb: 1 }}
          onChange={
            (e) => {
              setText(e.target.value);
            }
            //(link.current = { ...link.current, text: e.target.value })
          }
        />
        <Input
          type="url"
          autoFocus
          placeholder="https://example.com/"
          defaultValue={href}
          onChange={
            (e) => {
              setHref(e.target.value);
            }
            //(link.current = { ...link.current, href: e.target.value })
          }
        />
      </Flex>
      <Flex
        sx={{ justifyContent: "end" }}
        bg="var(--background-secondary)"
        p={1}
        px={2}
        mt={2}
      >
        <Button
          variant="dialog"
          onClick={() => {
            if (text !== "")
              editor
                .chain()
                .command(({ tr }) => {
                  tr.insertText(text, tr.mapping.map(from), tr.mapping.map(to));
                  tr.setSelection(
                    TextSelection.create(
                      tr.doc,
                      tr.mapping.map(from),
                      tr.mapping.map(to)
                    )
                  );
                  return true;
                })
                .focus()
                .run();

            editor
              .chain()
              .setMark("link")
              .setLink({ href, target: "_blank" })
              .run();

            setOpen(false);
          }}
        >
          Insert Link
        </Button>
      </Flex>
    </PopupWrapper>
  );
}

type linkPopup = {
  editor: Editor;
};

export function showLinkPopup(props: linkPopup) {
  showPopup({
    popup: () => (
      <LinkFloatingMenu
        onClose={() => console.log("onClose")}
        isOpen={true}
        editor={props.editor}
      />
    )
  });
}
