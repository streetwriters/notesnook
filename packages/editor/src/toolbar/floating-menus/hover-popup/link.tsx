/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { ToolbarGroup } from "../../components/toolbar-group.js";
import { HoverPopupProps } from "./index.js";
import { useHoverPopupContext } from "./context.js";
import { ToolbarGroupDefinition } from "../../types.js";

const LINK_TOOLS: ToolbarGroupDefinition = [
  "openLink",
  "editLink",
  "removeLink",
  "copyLink"
];
function LinkHoverPopup(props: HoverPopupProps) {
  const { editor } = props;
  const { selectedNode } = useHoverPopupContext();
  const { node } = selectedNode || {};

  if (
    !node?.isText ||
    node.marks.length <= 0 ||
    !node.marks.some((mark) => mark.type.name === "link")
  )
    return null;

  return (
    <ToolbarGroup
      force
      tools={LINK_TOOLS}
      groupId={"linkHoverTools"}
      editor={editor}
      sx={{
        bg: "background",
        boxShadow: "menu",
        borderRadius: "default",
        p: 1
      }}
    />
  );
}

export const LinkHoverPopupHandler = {
  isActive: (e: HTMLElement) => !!e.closest("a"),
  popup: LinkHoverPopup
};
