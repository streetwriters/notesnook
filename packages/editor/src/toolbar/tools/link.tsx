/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { ToolProps } from "../types.js";
import { ToolButton } from "../components/tool-button.js";
import { useRef, useState } from "react";
import { ResponsivePresenter } from "../../components/responsive/index.js";
import { LinkPopup } from "../popups/link-popup.js";
import { useToolbarLocation } from "../stores/toolbar-store.js";
import { MoreTools } from "../components/more-tools.js";
import { useRefValue } from "../../hooks/use-ref-value.js";
import { findMark, selectionToOffset } from "../../utils/prosemirror.js";
import { Flex, Link } from "@theme-ui/components";
import { ImageNode } from "../../extensions/image/index.js";
import { Link as LinkNode } from "../../extensions/link/index.js";
import { getMarkAttributes } from "@tiptap/core";
import { useHoverPopupContext } from "../floating-menus/hover-popup/context.js";
import { strings } from "@notesnook/intl";

export function LinkSettings(props: ToolProps) {
  const { editor } = props;
  const isBottom = useToolbarLocation() === "bottom";
  if (!editor.isActive("link") || !isBottom) return null;

  return (
    <MoreTools
      {...props}
      autoOpen
      autoCloseOnUnmount
      popupId="linkSettings"
      tools={
        editor.isEditable
          ? ["openLink", "editLink", "removeLink", "copyLink"]
          : ["openLink", "copyLink"]
      }
    />
  );
}

export function AddLink(props: ToolProps) {
  const { editor } = props;

  const isActive = editor.isActive("link");

  if (isActive) return <EditLink {...props} icon={"linkEdit"} />;
  return (
    <LinkTool
      {...props}
      onDone={(attributes) => editor.commands.toggleLink(attributes)}
      onClick={() => {
        if (!editor) return;
        const selectedText = editor.state.doc.textBetween(
          editor.state.selection.from,
          editor.state.selection.to
        );
        return { title: selectedText, href: "" };
      }}
    />
  );
}

export function AddInternalLink(props: ToolProps) {
  const { editor } = props;
  const isActive = editor.isActive(LinkNode.name);

  return (
    <ToolButton
      {...props}
      disabled={isActive}
      onClick={async () => {
        const link = await editor.storage.createInternalLink?.();
        if (!link) return;

        const selectedText = editor.state.doc.textBetween(
          editor.state.selection.from,
          editor.state.selection.to
        );
        editor.commands.setLink({ ...link, title: selectedText || link.title });
      }}
    />
  );
}

export function EditLink(props: ToolProps) {
  const { editor } = props;
  const { selectedNode: _selectedNode, hide } = useHoverPopupContext();
  const selectedNode = useRefValue(
    _selectedNode || selectionToOffset(editor.state)
  );
  const { node } = _selectedNode || {};
  const link = node ? findMark(node, LinkNode.name) : null;
  const attrs = link?.attrs || getMarkAttributes(editor.state, LinkNode.name);

  if (!editor.isEditable) return null;
  if (attrs && isInternalLink(attrs.href))
    return (
      <ToolButton
        {...props}
        onClick={async () => {
          hide();
          const link = await editor.storage.createInternalLink?.();
          if (!link) return;
          const { from, to } = editor.state.selection;
          if (selectedNode.current)
            editor.commands.setTextSelection(selectedNode.current);

          const selectedText =
            !!selectedNode.current &&
            editor.state.doc.textBetween(
              selectedNode.current.from,
              selectedNode.current.to
            );
          editor.commands.setLink({
            ...link,
            title: selectedText || link.title
          });
          if (selectedNode.current)
            editor.commands.setTextSelection({ from, to });
        }}
      />
    );

  return (
    <LinkTool
      {...props}
      isEditing
      onDone={(attributes) => {
        if (selectedNode.current)
          editor.chain().focus().setTextSelection(selectedNode.current).run();
        editor.commands.setLink(attributes);
        hide();
      }}
      onClick={() => {
        if (!selectedNode.current) return;

        const { node } = selectedNode.current;
        if (!node) return;

        const selectedText = node.textContent;
        const mark = findMark(node, "link");

        if (!mark) return;
        return {
          title: selectedText,
          href: mark.attrs.href
        };
      }}
    />
  );
}

export function RemoveLink(props: ToolProps) {
  const { editor } = props;
  const { selectedNode, hide } = useHoverPopupContext();
  if (!editor.isEditable) return null;
  return (
    <ToolButton
      {...props}
      toggled={false}
      onClick={() => {
        if (selectedNode)
          editor.commands.setTextSelection({
            from: selectedNode.from,
            to: selectedNode.to
          });
        editor.chain().focus().unsetLink().run();
        hide();
      }}
    />
  );
}

export function OpenLink(props: ToolProps) {
  const { editor } = props;
  const { selectedNode: _selectedNode, hide } = useHoverPopupContext();
  const selectedNode = useRefValue(
    _selectedNode || selectionToOffset(editor.state)
  );
  const { node } = selectedNode.current || {};
  const link = node ? findMark(node, "link") : null;
  if (!link) return null;
  const href = link?.attrs.href;

  return (
    <Flex sx={{ alignItems: "center" }}>
      <Link
        href={href}
        onClick={(e) => {
          e.preventDefault();
          editor.storage.openLink?.(href);
          hide();
        }}
        target="_blank"
        variant="body"
        sx={{
          fontSize: "subBody",
          fontFamily: "body",
          mr: 1,
          color: "accent",
          maxWidth: [150, 250],
          overflow: "hidden",
          textOverflow: "ellipsis",
          whiteSpace: "nowrap",
          ":visited": { color: "accent" },
          ":hover": { color: "accent", opacity: 0.8 }
        }}
      >
        {href}
      </Link>
      <ToolButton
        {...props}
        toggled={false}
        onClick={() => {
          editor.storage.openLink?.(href);
          hide();
        }}
      />
    </Flex>
  );
}

export function CopyLink(props: ToolProps) {
  const { editor } = props;
  const { selectedNode: _selectedNode, hide } = useHoverPopupContext();
  const selectedNode = useRefValue(
    _selectedNode || selectionToOffset(editor.state)
  );
  const { node } = selectedNode.current || {};
  const link = node ? findMark(node, "link") : null;
  if (!link) return null;
  const href = link?.attrs.href;

  return (
    <ToolButton
      {...props}
      toggled={false}
      onClick={() => {
        editor.storage.copyToClipboard?.(href);
        hide();
      }}
    />
  );
}

export type LinkDefinition = {
  href: string;
  title?: string;
};
type LinkToolProps = ToolProps & {
  isEditing?: boolean;
  onDone: (link: LinkDefinition) => void;
  onClick: () => LinkDefinition | undefined;
};
function LinkTool(props: LinkToolProps) {
  const { isEditing, onClick, onDone, editor, ...toolProps } = props;
  const buttonRef = useRef<HTMLButtonElement>(null);
  const [isOpen, setIsOpen] = useState(false);

  const [linkDefinition, setLinkDefinition] = useState<LinkDefinition>();
  const isImageActive = editor.isActive(ImageNode.name);

  return (
    <>
      <ToolButton
        {...toolProps}
        buttonRef={buttonRef}
        onClick={() => {
          const result = onClick();
          if (!result) return;
          setLinkDefinition(result);
          setIsOpen(true);
        }}
        toggled={isOpen}
      />
      <ResponsivePresenter
        mobile="sheet"
        desktop="popup"
        position={{
          target: buttonRef.current || undefined,
          isTargetAbsolute: true,
          location: "below",
          align: "center",
          yOffset: 5
        }}
        title={isEditing ? strings.editLink() : strings.insertLink()}
        isOpen={isOpen}
        items={[]}
        onClose={() => {
          setIsOpen(false);
          editor.commands.focus();
        }}
        focusOnRender={false}
      >
        <LinkPopup
          link={linkDefinition}
          isEditing={isEditing}
          isImageActive={isImageActive}
          onClose={() => setIsOpen(false)}
          onDone={(link) => {
            onDone(link);
            setIsOpen(false);
          }}
        />
      </ResponsivePresenter>
    </>
  );
}

export function isInternalLink(href?: string | null) {
  return typeof href === "string" ? href.startsWith("nn://") : false;
}
