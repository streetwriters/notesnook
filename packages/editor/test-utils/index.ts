/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { Editor, AnyExtension, Extensions } from "@tiptap/core";
import StarterKit from "@tiptap/starter-kit";
import { builders, NodeBuilder } from "prosemirror-test-builder";
import { Schema } from "@tiptap/pm/model";

type Builder<TNodes extends string> = {
  scheme: Schema;
} & Record<TNodes, NodeBuilder>;

type EditorOptions<TNodes extends string> = {
  element?: HTMLElement;
  extensions: Record<TNodes, AnyExtension | false>;
  initialContent?: string;
};

export function createEditor<TNodes extends string>(
  options: EditorOptions<TNodes>
) {
  const { extensions, initialContent, element } = options;
  const editor = new Editor({
    element,
    content: initialContent,
    extensions: [
      StarterKit.configure({
        ...Object.entries(extensions).reduce(
          (prev, [name]) => ({
            ...prev,
            [name]: false
          }),
          {}
        )
      }),
      ...(Object.values(extensions) as Extensions)
    ]
  });

  const builder = builders(editor.schema) as unknown as Builder<TNodes>;

  return { editor, builder };
}

export function h<K extends keyof HTMLElementTagNameMap>(
  tag: K,
  children: (HTMLElement | string)[] = [],
  attr: Record<string, string | undefined> = {}
): HTMLElementTagNameMap[K] {
  const element = document.createElement(tag);
  element.append(
    ...children.map((v) =>
      typeof v === "string" ? document.createTextNode(v) : v
    )
  );
  for (const key in attr) {
    const value = attr[key];
    if (value) element.setAttribute(key, value);
  }
  return element;
}

function elem<K extends keyof HTMLElementTagNameMap>(tag: K) {
  return function (
    children: (HTMLElement | string)[] = [],
    attr: Record<string, string | undefined> = {}
  ): HTMLElementTagNameMap[K] {
    return h(tag, children, attr);
  };
}

export const ul = elem("ul");
export const li = elem("li");
export const p = elem("p");

export function text(text: string) {
  return document.createTextNode(text);
}
