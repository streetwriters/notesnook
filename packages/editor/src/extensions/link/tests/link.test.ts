/*
This file is part of the Notesnook project (https://notesnook.com/)

Copyright (C) 2023 Streetwriters (Private) Limited

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

import { describe, expect, test } from "vitest";
import { createEditor, h } from "../../../../test-utils";
import { Link, linkRegex } from "../link";

const input = [
  {
    text: "[Example](https://www.google.com)",
    expected: ["[Example](https://www.google.com)"]
  },
  {
    text: "[Example](https://www.example1.com) and [Example2](https://www.example2.com)",
    expected: [
      "[Example](https://www.example1.com)",
      "[Example2](https://www.example2.com)"
    ]
  },
  {
    text: "[]()",
    expected: null
  },
  {
    text: "[] [Link](http://example.com)",
    expected: ["[Link](http://example.com)"]
  },
  {
    text: "[not a link] [Link](http://example.com)",
    expected: ["[Link](http://example.com)"]
  },
  {
    text: "[NoLink]",
    expected: null
  }
];

input.forEach(({ text, expected }, i) => {
  test(`link regex ${text}`, () => {
    const result = text.match(linkRegex);
    expect(result).toEqual(expected);
  });
});

describe("paste text", () => {
  test("with markdown link", async () => {
    const editorElement = h("div");
    const { editor } = createEditor({
      element: editorElement,
      extensions: {
        link: Link
      }
    });

    const clipboardEvent = new Event("paste", {
      bubbles: true,
      cancelable: true,
      composed: true
    });

    (clipboardEvent as unknown as any)["clipboardData"] = {
      getData: (type: string) =>
        type === "text/plain" ? "[test](example.com)" : undefined
    };

    editor.view.dom.dispatchEvent(clipboardEvent);

    await new Promise((resolve) => setTimeout(resolve, 100));

    expect(editorElement.outerHTML).toMatchSnapshot();
  });

  test("with multiple markdown links", async () => {
    const editorElement = h("div");
    const { editor } = createEditor({
      element: editorElement,
      extensions: {
        link: Link
      }
    });

    const clipboardEvent = new Event("paste", {
      bubbles: true,
      cancelable: true,
      composed: true
    });

    (clipboardEvent as unknown as any)["clipboardData"] = {
      getData: (type: string) =>
        type === "text/plain"
          ? "[test](example.com) some text [test2](example2.com)"
          : undefined
    };

    editor.view.dom.dispatchEvent(clipboardEvent);

    await new Promise((resolve) => setTimeout(resolve, 100));

    expect(editorElement.outerHTML).toMatchSnapshot();
  });
});
