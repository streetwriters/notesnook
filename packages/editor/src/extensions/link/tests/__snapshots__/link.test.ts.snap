// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`paste text > with link 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><a target="_blank" rel="noopener noreferrer nofollow" href="http://example.com">example.com</a></p></div></div>"`;

exports[`paste text > with markdown link 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><a target="_blank" rel="noopener noreferrer nofollow" href="example.com">test</a></p></div></div>"`;

exports[`paste text > with multiple links 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><a target="_blank" rel="noopener noreferrer nofollow" href="http://example.com">example.com</a> <a target="_blank" rel="noopener noreferrer nofollow" href="http://example2.com">example2.com</a></p></div></div>"`;

exports[`paste text > with multiple markdown links 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><a target="_blank" rel="noopener noreferrer nofollow" href="example.com">test</a> some text <a target="_blank" rel="noopener noreferrer nofollow" href="example2.com">test2</a></p></div></div>"`;

exports[`unformatted paste text > with link shouldn't format as link 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p>https://example.com</p></div></div>"`;

exports[`unformatted paste text > with markdown link should still format as link 1`] = `"<div><div contenteditable="true" translate="no" class="tiptap ProseMirror" tabindex="0"><p><a target="_blank" rel="noopener noreferrer nofollow" href="example.com">asdf</a></p></div></div>"`;
